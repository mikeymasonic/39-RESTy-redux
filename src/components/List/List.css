.List {
  background-color: purple;
  color: whitesmoke;
  padding: 20px;
  margin: 20px;

  ul {
    list-style-type: none;
    padding-inline-start: 0;
  }

  li {
    background-color: grey;
    color: black;
    margin: 10px;
    width: 400px;
  }

  li:hover {
    background-color: purple;
    color: whitesmoke;
  }

  h3 {
    margin-block-start: 0;
    margin-block-end: 0;
  }

  p {
    margin-block-start: 0;
    margin-block-end: 0;
  }
}
