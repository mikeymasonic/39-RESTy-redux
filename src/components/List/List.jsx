import React from 'react';
import PropTypes from 'prop-types';
import styles from './List.css';
import { getRequests } from '../../selectors/restySelectors';
import { useSelector } from 'react-redux';

const List = ({ handleClear, handleLoad }) => {
  const requests = useSelector(getRequests);
  const requestNodes = requests.map(request => {
    const home = request.url.split('/')[2];
    const user = request.url.split('.com')[1];
    return <li key={`${request.url}${request.method}`} onClick={() => handleLoad(request.url, request.method, request.body)}>
      <h3>{request.method}</h3>
      <p>{home}</p>
      <p>{user}</p>
    </li>;
  });

  return (
    <div className={styles.List}>
      <ul>
        {requestNodes}
      </ul>
      <button onClick={handleClear}>Clear</button>
    </div>
  );
};

List.propTypes = {
  requests: PropTypes.array,
  handleClear: PropTypes.func,
  handleLoad: PropTypes.func
};

export default List;
