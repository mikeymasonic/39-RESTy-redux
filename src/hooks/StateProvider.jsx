import React, { createContext, useReducer, useContext } from 'react';
import PropTypes from 'prop-types';

const StateContext = createContext();

const initialState = {
  url: '',
  method: 'GET',
  body: '',
  disable: true,
  headers: {},
  response: {},
  requests: []
};

export function reducer(state, action) {
  switch(action.type) {
    case 'SET_URL':
      return { ...state, url: action.payload };
    case 'SET_METHOD':
      return { ...state, method: action.payload };
    case 'SET_BODY':
      return { ...state, body: action.payload };
    case 'SET_REQUESTS':
      return { ...state, requests: action.payload };
    case 'SET_DISABLE':
      return { ...state, disable: action.payload };
    case 'SET_RESPONSE':
      return { ...state, response: action.payload };
    case 'SET_HEADERS':
      return { ...state, headers: action.payload };
    
    default:
      return state;
  }
}

export const StateProvider = ({ children }) => {
  const [state, dispatch] = useReducer(reducer, initialState);

  return (
    <StateContext.Provider value={{ state, dispatch }}>
      {children}
    </StateContext.Provider>
  );
};

StateProvider.propTypes = {
  children: PropTypes.node
};

export const useGlobalState = () => {
  const { state } = useContext(StateContext);
  return state;
};

export const useURL = () => {
  const { url } = useGlobalState();
  return url;
};

export const useBody = () => {
  const { body } = useGlobalState();
  return body;
};

export const useMethod = () => {
  const { method } = useGlobalState();
  return method;
};

export const useDispatch = () => {
  const { dispatch } = useContext(StateContext);
  return dispatch;
};

export const useDisable = () => {
  const { disable } = useGlobalState();
  return disable;
};

export const useResponse = () => {
  const { response } = useGlobalState();
  return response;
};

export const useHeaders = () => {
  const { headers } = useGlobalState();
  return headers;
};

export const useRequests = () => {
  const { requests } = useGlobalState();
  return requests;
};
