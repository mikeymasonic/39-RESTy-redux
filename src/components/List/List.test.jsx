import React from 'react';
import { shallow } from 'enzyme';
import List from './List';
import { StateProvider } from '../../hooks/StateProvider';

describe('List component', () => {
  it('renders List', () => {
    const wrapper = shallow(
      <StateProvider>
        <List />
      </StateProvider>
    );
    expect(wrapper).toMatchSnapshot();
  });
});
