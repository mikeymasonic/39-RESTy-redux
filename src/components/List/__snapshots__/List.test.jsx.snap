// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`List component renders List 1`] = `
<ContextProvider
  value={
    Object {
      "dispatch": [Function],
      "state": Object {
        "body": "",
        "disable": true,
        "headers": Object {},
        "method": "GET",
        "requests": Array [],
        "response": Object {},
        "url": "",
      },
    }
  }
>
  <List />
</ContextProvider>
`;
